### PR TITLE
Marked Debug implementations for primitive types as #[inline]

### LINKED ISSUE #47792: Mark Debug implementations for primitive types as #[inline]
This should allow them to forward directly to the Display implementations without generating adapters.

See https://github.com/rust-lang/rust/issues/47785#issuecomment-360896774

### PR DESCRIPTION
**Pull request by ghost**
Fixes #47792.

### CODE REVIEW

File: src/libcore/fmt/num.rs [hunk 1]
```diff
@@ -157,6 +157,7 @@ macro_rules! debug {
     ($T:ident) => {
         #[stable(feature = "rust1", since = "1.0.0")]
         impl fmt::Debug for $T {
+            #[inline]
```
Comment (bluss) [inline src/libcore/fmt/num.rs]: This one, definitely

File: src/libcore/fmt/mod.rs [hunk 2]
```diff
@@ -1718,6 +1725,7 @@ macro_rules! tuple {
         #[stable(feature = "rust1", since = "1.0.0")]
         impl<$($name:Debug),*> Debug for ($($name,)*) where last_type!($($name,)+): ?Sized {
             #[allow(non_snake_case, unused_assignments, deprecated)]
+            #[inline]
             fn fmt(&self, f: &mut Formatter) -> Result {
```
Comment (bluss) [inline src/libcore/fmt/mod.rs]: This is an example of one we don't need to mark inline. We want to mark those that are not generic and forwarding directly to another non-generic function as inline.

File: src/libcore/fmt/mod.rs [hunk 3]
```diff
@@ -1628,6 +1632,7 @@ impl Display for str {
 
 #[stable(feature = "rust1", since = "1.0.0")]
 impl Debug for char {
+    #[inline]
     fn fmt(&self, f: &mut Formatter) -> Result {
```
Comment (bluss) [inline src/libcore/fmt/mod.rs]: For example, this one should not be inline.

**Comment by kennytm** [comment]
This PR caused a linker error in the `ui/nll/constant.rs` test...

File: src/libcore/fmt/mod.rs (hunk 2 above)
Comment (penpalperson) [reply src/libcore/fmt/mod.rs]: Thanks! With those conditions, should the only ones that I mark inline be non-float numbers, the unit type and bools? Just making sure.

Comment (bluss) [reply src/libcore/fmt/mod.rs]: Yes, I agree.

I'm not 100% sure which trade off is the best for the function pointers, ideas welcome.

Comment (bluss) [reply src/libcore/fmt/mod.rs]: I'd even skip the unit type if pad is a method that inlines. We don't want to inline a bigger function with this PR.

Comment (penpalperson) [reply src/libcore/fmt/mod.rs]: The unit type's pad method doesn't inline, so I think I'll inline the unit type's fmt function. For function pointers, I think I'll avoid marking them as inline them for now.

**Comment by bluss** [comment]
Thanks a lot!

@bors r+ rollup

**Comment by bluss** [comment]
Before bors gets around to merging this, you can change the PR message (first comment in this issue) to use one of the command forms for closing the issue [that github likes](https://help.github.com/articles/closing-issues-using-keywords/)

**Pull request merged**